tests/collaborative/collaborative_monkey_party.test.ts

```diff
@@ -0,0 +1,310 @@
+  isTargetDependent,
+  UnboundedZone,
+} from "../../src/types";
+import { getRangeValues } from "../test_helpers";
+import { MockTransportService } from "../__mocks__/transport_service";
+import { setupCollaborativeEnv } from "./collaborative_helpers";
+import { getMonkeyPartyCommands } from "./revisions_party";
+
+/**
+ * The monkey party test simulates random user actions in a collaborative environment
+ * to ensure all users remain synchronized.
+ *
+ * If a randomly generated test fails, it minimizes the failing commands and
+ * generates a minimal test case that you can copy-paste from the console.
+ *
+ * By default, this test runs once with a deterministic seed to prevent
+ * non-deterministic behavior in the CI pipeline.
+ *
+ * It is intended to be run locally from time to time by increasing `PARTY_COUNT`.
+ */
+
+const PARTY_COUNT = 1; // increase this number to run more tests and enjoy the parties 🐵🎉
+
+describe("monkey party", () => {
+  const now = Date.now();
+  const seeds = range(0, PARTY_COUNT).map((i) => (now + i).toString());
+
+  test.each(PARTY_COUNT === 1 ? ["deterministic-seed"] : seeds)(
+    "monkey party with seed %s",
+    (seed) => {
+      seedrandom(seed, { global: true });
+      const { network, alice, bob, charlie } = setupCollaborativeEnv();
+      const commands = getMonkeyPartyCommands();
+      addUndoRedo(commands);
+      randomizeCommandsPayload(commands);
+
+      const actions = assignUser(shuffle(commands), [alice, bob, charlie]);
+      const concurrentActions = randomizeConcurrentActions(actions);
+      const { fail, executedActions } = run(network, [alice, bob, charlie], concurrentActions);
+      if (fail) {
+        const generatedTestCode = actionsToTestCode(
+          `failed with seed ${seed}`,
+          minimizeFailingCommands(executedActions)
+        );
+        console.log(generatedTestCode);
+        throw new Error(`Failed with seed ${seed}!`);
+      }
+      expect([alice, bob, charlie]).toHaveSynchronizedExportedData();
+    }
+  );
+
+  test("missing core commands", (done) => {
+    const missingCommands: Set<string> = new Set(coreTypes);
+
+    for (const command of getMonkeyPartyCommands()) {
+      missingCommands.delete(command.type);
+    }
+
+    // Commands that are never dispatched by the user. They are only sub-commands
+    // dispatched by the core plugins.
+    const INTERNAL_CORE_COMMANDS = ["UPDATE_CELL_POSITION", "CREATE_FIGURE"];
+    for (const command of INTERNAL_CORE_COMMANDS) {
+      missingCommands.delete(command);
+    }
+    if (missingCommands.size) {
+      const failMessage = `Missing core commands for the monkey party test! 🐵🎉
+Monkeys need all available commands to have fun.
+
+Please add the following missing commands to \`revision_party.ts\`:
+
+${Array.from(missingCommands).join("\n")}`;
+      done(failMessage);
+    } else {
+      done();
+    }
+  });
+});
+
+function addUndoRedo(commands: CoreCommand[], undoRedoRatio = 0.15) {
```

Comment:
argument ratio is never used

Reply:
They could be included in `getMonkeyPartyCommands` if the ratio never changes. Or directly in a random command generator 😛